### PR TITLE
Add/guard installation flow

### LINKED ISSUE #58297: Plugin Installation: Better guard the installation flow
During our demo it was suggested that we should better guard our plugin installation direct url so that we avoid malicious users getting WordPress.com users to visit it and thus install a malicious wporg plugin in their site.

### PR DESCRIPTION
**Pull request by gcsecsey**
#### Changes proposed in this Pull Request

- [x] prevents the installation if the URL is visited without clicking the Install button on the plugins page
- [x] if plugin `A` was installed previously, it still prevents the installation of plugin `B` from the URL directly

#### Prevents direct access
1.  choose a site with any plan
2. visit `/marketplace/<pluginname>/install/<siteURL>`
3. verify that the this error message is shown:
image

#### Allows install after `Install` button click
1.  choose a site with any plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Install` or `Upgrade and Install` button
4. perform the upgrade if needed
5. verify that the install executes normally

#### Throws incorrect plan error message correctly
1.  choose a site with a free plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Upgrade and Install` button
4. from the checkout page, click the X button in the upper left corner
5. verify that the this error message is shown:
image

#### Prevents direct access after a successful plugin install
1.  choose a site with any plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Install` or `Upgrade and Install` button
4. perform the upgrade if needed
5. verify that the install executes normally
6. visit `/marketplace/<secondpluginname>/install/<siteURL>`
7. verify that the this error message is shown:
image

#### Prevents direct access
1.  choose a site with any plan
2. visit `/marketplace/install/<siteURL>`
3. verify that the this error message is shown:
image

#### Allows install after plugin upload
1.  choose a site with any plan
2. visit `/plugins/upload/<siteURL>`
3. upload a plugin
4. verify that the install executes normally

Fixes #58297

### CODE REVIEW

**Comment by gcsecsey** [comment]
When displaying the error pages, there's a `margin-top: 350px` rule applied on the container:
image

@cpapazoglou @WBerredo any ideas why it's needed or if we can remove it perhaps?

**Comment by cpapazoglou** [comment]
> When displaying the error pages, there's a `margin-top: 350px` rule applied on the container: image
> 
> @cpapazoglou @WBerredo any ideas why it's needed or if we can remove it perhaps?

I suppose it was developed that it appears vertically centered? Probably we can make it better by utilizing `flex`?

SS 2021-11-25 at 13 55 28

**Comment by gcsecsey** [comment]
As we're handling multiple types of errors now in this component, I've tried consolidating the error handling with the help of an enum, but ran into some race condition issues down the line.

I see we're checking some conditions in an async way here: https://github.com/Automattic/wp-calypso/blob/0d38fdc2df4bfeb26135270a69f6d6e0e83cc8c5/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L98

@WBerredo could you help me with why the callback is needed here? Do we have some piece of state that changes on page load? Thanks 🙂

**Comment by WBerredo** [comment]
This `useEffect` will run at every render (we don't pass the properties array) and check there is an unsupported plan, if there is, it will check again in 2 seconds and if the problem persists, it shows an error.
https://github.com/Automattic/wp-calypso/blob/0d38fdc2df4bfeb26135270a69f6d6e0e83cc8c5/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L96-L102


> why the callback is needed here?

On the **upgrade and install** flow, the plan we receive on the first render is the old/low tier plan, but this plan gets updated on the subsequent renders. That's why we wait 2 seconds before showing the error, and we consider if it didn't get updated in 2 seconds this is the current plan of the user.

---

Does that makes sense? There is anything that still needs explaining?

**Comment by gcsecsey** [comment]
@WBerredo I thought that we can't use the data that we have on the initial render, but wasn't sure, thanks for clarifying! 🎉

**Review by WBerredo — COMMENTED** [review]
I was able to test the general path, but I would like to point two things:

* I didn't get the "upgrade to business plan" error, when following the steps I got the "direct access" error.
* Shouldn't we store which plugin triggered the install and check it on the state? Otherwise, I would be able to trigger an install and access the direct URL with another plugin installation if done at the same time.

**Comment by gcsecsey** [comment]
Thanks for testing this @WBerredo, it'd be a great idea to add the plugin to the state! I'll check the issue with the errors and add this additional check. 🙂

File: client/my-sites/plugins/plugin-details.jsx [hunk 1]
```diff
@@ -436,6 +441,15 @@ function onClickInstallPlugin( { dispatch, selectedSite, slug, upgradeAndInstall
...
 
+	dispatch( productToBeInstalled( null, slug, selectedSite.slug ) );
+
+	dispatch(
+		pluginInstallationStateChange(
+			MARKETPLACE_ASYNC_PROCESS_STATUS.IN_PROGRESS,
+			'preauthorize plugin installation URL'
+		)
+	);
+
```
Comment (cpapazoglou) [inline client/my-sites/plugins/plugin-details.jsx]: We will need to add similar events to the Plugin Upload flow `https://wordpress.com/plugins/upload/[DOMAIN]` which is now returning an error. 

https://github.com/Automattic/wp-calypso/blob/08961f239b9c353bb0b44983d9f7b014db038133/client/my-sites/plugins/plugin-upload/index.jsx#L60-L62

Comment (gcsecsey) [reply client/my-sites/plugins/plugin-details.jsx]: Thanks @cpapazoglou, that's a very good point!

**Comment by gcsecsey** [comment]
> Shouldn't we store which plugin triggered the install and check it on the state? Otherwise, I would be able to trigger an install and access the direct URL with another plugin installation if done at the same time.

>We will need to add similar events to the Plugin Upload flow https://wordpress.com/plugins/upload/[DOMAIN] which is now returning an error.

I've added these changes and updated the test instructions with steps for the plugin upload flow. When trying to visit the install page directly on https://wordpress.com/plugins/upload/[DOMAIN], we're now showing a flow-specific error message that helps the user navigate back to the plugin upload page.

> I didn't get the "upgrade to business plan" error, when following the steps I got the "direct access" error.

This was a small mistake on my side, I've accidentally copied over the other error message. 😄 I've fixed it now, it should work as per the test instructions. 👍

**Review by WBerredo — CHANGES_REQUESTED** [review]
I am seeing an error on the **upgrade and install flow**, if I complete the purchase or not I am always getting the "This URL should not be accessed directly. Please click the Install button on the plugin page." error.

I'm not sure if this is related to the environment it is created on the PR. But I think this worth some research.

**Comment by gcsecsey** [comment]
> I am seeing an error on the upgrade and install flow, if I complete the purchase or not I am always getting the "This URL should not be accessed directly. Please click the Install button on the plugin page." error.

I've looked into this, and found that previously, the `pluginInstallationStatus` state var was written from both actions that were fired, resulting in a race condition when persisting the state:
https://github.com/Automattic/wp-calypso/blob/91a7d3ba948945bff26f626566ce1de335339a61/client/my-sites/plugins/plugin-details.jsx#L444-L451

https://github.com/Automattic/wp-calypso/blob/a184952a4afc7f4c61b1c9f395493576b6d387fd/client/state/marketplace/purchase-flow/reducer.ts#L34-L38

I've updated the installation check to still validate the `primaryDomain` and `productSlug` variables, and to check that the plugin installation hasn't been completed yet:
https://github.com/Automattic/wp-calypso/blob/a184952a4afc7f4c61b1c9f395493576b6d387fd/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L82

https://github.com/Automattic/wp-calypso/blob/a184952a4afc7f4c61b1c9f395493576b6d387fd/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L87

**Comment by gcsecsey** [comment]
With @WBerredo we've discovered a new edge case.

When the user navigates away from the checkout page by changing the URL in the browser, the IndexedDB state is reset to the `defaultState` value:
https://d.pr/v/rDDfH9

However, when the user navigates away from the checkout page by clicking the X icon in the upper left corner of the checkout page, the IndexedDB state is persisted as expected:
https://d.pr/v/tqbGqp

It seems that by navigating to a new URL, the Redux init clears the persisted state instead of loading the values from IndexedDB, and hence breaks the purchase flow. When navigating through client routes, the `defaultState` isn't loaded again and the feature works as expected.

We've tried removing the `defaultState` so that values in IndexedDB can be read back on a later visit. However, this results in an error, the slice reducer can't be initiated without a default state:
image

It seems that we won't be able to persist state between URL hops by utilizing IndexedDB. For the time being, I've updated the testing steps to clarify that the feature only works when navigating via UI elements.

**Review by WBerredo — APPROVED** [review]
Tested the last version and it is all working good 🚀

PS: The only edge case I found is mentioned at this PR but I don't think we can do anything about it.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7024500

Thank you @gcsecsey for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7024500).

**Pull request merged**